### PR TITLE
Refactor binary search logic for ValidatorsAccumulator.sol

### PR DESCRIPTION
**Pull request by Owliie**
Remove validators count and therefore refactor the storage layout and storage reading

### CODE REVIEW

File: beacon-light-client/solidity/contracts/validators_accumulator/ValidatorsAccumulator.sol [hunk 1]
```diff
@@ -81,28 +80,31 @@ contract ValidatorsAccumulator is IValidatorsAccumulator {
...
-    });
+    uint64 blockNumber = uint64(block.number);
+    snapshots[blockNumber] = _getRoot(validatorsCount);
+    uint256 blockNumbersLength = blockNumbers.length;
+    if (
+      blockNumbersLength == 0 ||
+      blockNumbers[blockNumbersLength - 1] != blockNumber
+    ) {
+      blockNumbers.push(blockNumber);
+    }
```
Comment (Xearty) [inline beacon-light-client/solidity/contracts/validators_accumulator/ValidatorsAccumulator.sol:91]: I think it would be better if we used a ring buffer

File: beacon-light-client/solidity/contracts/validators_accumulator/ValidatorsAccumulator.sol [hunk 2]
```diff
@@ -81,28 +80,31 @@ contract ValidatorsAccumulator is IValidatorsAccumulator {
       size /= 2;
     }
 
-    snapshots[validatorsCount - 1] = DepositData({
-      blockNumber: block.number,
-      accumulator: _getRoot(validatorsCount)
-    });
+    uint64 blockNumber = uint64(block.number);
+    snapshots[blockNumber] = _getRoot(validatorsCount);
```
Comment (Xearty) [inline beacon-light-client/solidity/contracts/validators_accumulator/ValidatorsAccumulator.sol:84]: Is there a reason for not using getValidatorsAccumulator?

File: beacon-light-client/solidity/contracts/validators_accumulator/ValidatorsAccumulator.sol (hunk 1 above)
Comment (Dimo99) [reply beacon-light-client/solidity/contracts/validators_accumulator/ValidatorsAccumulator.sol:91]: We have discussed the idea of the ring buffer but because it is a very permissionless protocol we decided that someone could make multiple deposits and delete old data, maybe this assumption is extremely pessimistic as those deposits are not cheap and we learned later the contract is not entirely permissionless. 
But apart from the ring buffer, I like the idea of using an array actually instead of a map I think it will make the solution a bit cleaner
```
DepositData[] snapshots;
```
And if the deposit is for the same block we will just overwrite the deposit

File: beacon-light-client/solidity/contracts/validators_accumulator/ValidatorsAccumulator.sol [hunk 3]
```diff
@@ -15,10 +15,9 @@ contract ValidatorsAccumulator is IValidatorsAccumulator {
 
   // A counter for the total number of validators
   uint256 internal validatorsCount;
-  // Start index of validators map
-  uint256 internal startIndex;
```
Comment (Dimo99) [inline beacon-light-client/solidity/contracts/validators_accumulator/ValidatorsAccumulator.sol:19]: We should do that in another PR as it was reported as separate issue by audtiors

File: beacon-light-client/solidity/contracts/validators_accumulator/ValidatorsAccumulator.sol (hunk 1 above)
Comment (Owliie) [reply beacon-light-client/solidity/contracts/validators_accumulator/ValidatorsAccumulator.sol:91]: Using mapping + array costs less in both `deposit` and `verify` transactions. Therefore, we will stick to the current approach.

File: beacon-light-client/solidity/contracts/validators_accumulator/ValidatorsAccumulator.sol [hunk 4]
```diff
@@ -125,32 +123,34 @@ contract ValidatorsAccumulator is IValidatorsAccumulator {
...
+    uint64 upperBlockNumber = blockNumbers[upper];
+    if (upperBlockNumber <= blockNumber) {
+      return upperBlockNumber;
     }
 
-    if (snapshots[lower].blockNumber > blockNumber) {
-      return 0;
+    uint64 lowerBlockNumber = blockNumbers[lower];
+    if (lowerBlockNumber > blockNumber) {
+      return lowerBlockNumber;
```
Comment (Dimo99) [inline beacon-light-client/solidity/contracts/validators_accumulator/ValidatorsAccumulator.sol:138]: In my opinion, it is better to return type(uint64).max as this is a special case where no block was found. And will make the code a bit easier to follow

**Review by Xearty — APPROVED** [review]
